### PR TITLE
WIP: Add support for detecting tests in source files, and implement it for Rust

### PR DESCRIPTION
**Pull request by mikayla-maki**
TODO:
- [x] Wire up a new query
- [x] Use this query to find tests in the current file range
- [x] Render a play button in that location
- [ ] Spawn the test template from that play button when clicked

Release Notes:

- TODO

### CODE REVIEW

File: crates/languages/src/rust/config.toml [hunk 1]
```diff
@@ -12,3 +12,6 @@ brackets = [
     { start = "/*", end = " */", close = true, newline = false, not_in = ["string", "comment"] },
 ]
 collapsed_placeholder = " /* ... */ "
+test_templates = [
+    {capture_name = "cargo-test", label = "cargo test", template = "cargo test -p ZED_PACKAGE ZED_TEST_CAPTURES[0]"}
```
Comment (osiewicz) [inline crates/languages/src/rust/config.toml:16]:
```suggestion
    {capture_name = "cargo-test", label = "cargo test", template = "cargo test -p $ZED_PACKAGE $ZED_TEST_CAPTURES[0]"}
```
these variables are env variables proper.
I wonder if `shellexpand` supports arrays 🤔 We use that library to expand uses of Zed environment variables in labels. If it does support it, we could put the name of a test in a label and the task system will expand it for us.

**Comment by osiewicz** [comment]
A question/food for thought: will all of the tests show up in a task modal?
On one hand I think that'll bloat the modal, but on the other it would be inconsistent to not include them.

**Comment by davidbarsky** [comment]
Hi! Sorry for butting in (asking since I'm a maintainer on rust-analyzer and I use Zed...), but would it be possible to use [rust-analyzer's runnable](https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#runnables) functionality instead, by any chance?

(We'll be changing to extend support for non-Cargo build systems in https://github.com/rust-lang/rust-analyzer/pull/16840, but we don't expect to change the custom LSP request beyond what's in that PR.)

**Comment by osiewicz** [comment]
Hey,
We did consider using an lsp extension for tasks (see https://github.com/zed-industries/zed/issues/7108), but it does not solve test discoverability problem for language servers without these kinds of protocol extensions. In theory though we can consume VSC tasks just fine.

File: crates/languages/src/rust/config.toml (hunk 1 above)
Comment (osiewicz) [reply crates/languages/src/rust/config.toml:16]: Ok, I've checked with shellexpand and we can use it to substitute array subscripts just fine, e.g. during parsing of the command we'll come across `ZED_TEST_CAPTURES[0]`; we have to parse the subscript part ourselves, but that shouldn't be too hard :)

**Comment by davidbarsky** [comment]
> Hey,
> 
> We did consider using an lsp extension for tasks (see https://github.com/zed-industries/zed/issues/7108), but it does not solve test discoverability problem for language servers without these kinds of protocol extensions. In theory though we can consume VSC tasks just fine.

Thanks! I read through that, I appreciate you linking to your thinking on the matter! Obviously, y'all are the maintainers, but I'm curious as to how feasible it'd be to take a treesitter-based approach for _all_ languages as a baseline and enhance specific languages (such as Rust) with LSP extensions. The benefit of LSP-based tasks would be more accurate task generation because rust-analyzer is able to see through macro expansions and whatnot. 

From an implementation standpoint, I think there might be some complications around scaling this approach/figuring out how it interacts with Zed's extension API, but for rust-analyzer at least, I'd be happy to send PRs to Zed each time that request corresponding to runnables changes (the PR I linked to would be first change in 3-4 years, fwiw). 

I don't expect much would change on the Zed side of things, since I think the task templates y'all have are sufficiently flexible. I _think_ the integration boils down to mapping the right fields in the runnable response to the pre-existing task template in Zed.

File: crates/languages/src/rust/config.toml (hunk 1 above)
Comment (mikayla-maki) [reply crates/languages/src/rust/config.toml:16]: This was discussed further in a pairing session, @osiewicz and I agreed that putting the tasks in the config file is unnecessary, we can just expose the test capture names and make a linking in the existing tasks.json file.

**Comment by davidbarsky** [comment]
> > but I'm curious as to how feasible it'd be to take a treesitter-based approach for all languages as a baseline and enhance specific languages (such as Rust) with LSP extensions.
> 
> That's where I'm thinking we'll end up :D We'd love support on maintaining experimental API protocols like this :) @davidbarsky

That's great to hear! I wasn't sure how open y'all would be to that, but I'm very happy you are! I'll keep my eyes peeled on opportunities for changes that make that possible.

**Comment by osiewicz** [comment]
Closing this in favor of #11195

**Pull request closed without merging**